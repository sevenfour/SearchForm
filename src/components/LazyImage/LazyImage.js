import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { Landscape } from '../icons';

import styles from './LazyImage.module.css';

const placeholderSrc = (width, height) =>
  `data:image/svg+xml,%3Csvg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${width} ${height}"%3E%3C/svg%3E`;

// NOTE: the overal image lazy loading can be significantly imporved further

const LazyImage = props => {

  const {
    alt,
    mainUrl,
    width,
    height,
    ...rest
  } = props;

  const [isError, setIsError] = useState(false);
  const [imageSrc, setImageSrc] = useState(placeholderSrc(width, height));

  function handleLoad() {
    setImageSrc(mainUrl);
  }

  function handleError() {
    setIsError(true);
  }  

  return mainUrl && !isError
    ? (
      <img
        alt={alt}
        className={styles.image}
        onError={handleError}
        onLoad={handleLoad}
        src={imageSrc}
        {...rest}
      />
    )
    : <Landscape className={styles.placeholderImage} />;
};

LazyImage.propTypes = {
  alt: PropTypes.string,
  height: PropTypes.string,
  mainUrl: PropTypes.string,
  smallUrl: PropTypes.string,
  width: PropTypes.string
};

export default  LazyImage;
