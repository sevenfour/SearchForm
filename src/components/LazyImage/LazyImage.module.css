
.image {
  max-width: 100%;
  height: auto;
}

.placeholderImage {
  width: 100%;
  height: 200px;
  color: var(--stone-grey);
}
