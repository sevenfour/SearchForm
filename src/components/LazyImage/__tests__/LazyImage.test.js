import React from 'react';
import { shallow } from 'enzyme';

import LazyImage from '..';

describe('<LazyImage /> component tests', () => {

  it('renders with a placeholder', () => {
    const lazyImage = shallow(
      <LazyImage />
    );

    expect(lazyImage).toMatchSnapshot();
  });

  it('renders with an image', () => {
    const lazyImage = shallow(
      <LazyImage mainUrl="/qwerty" />
    );

    expect(lazyImage).toMatchSnapshot();
  });

});
