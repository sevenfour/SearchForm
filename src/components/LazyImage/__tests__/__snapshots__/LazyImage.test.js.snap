// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LazyImage /> component tests renders with a placeholder 1`] = `
<Landscape
  className="placeholderImage"
/>
`;

exports[`<LazyImage /> component tests renders with an image 1`] = `
<img
  className="image"
  onError={[Function]}
  onLoad={[Function]}
  src="data:image/svg+xml,%3Csvg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 undefined undefined\\"%3E%3C/svg%3E"
/>
`;
